[Mime] Fix undefined array key 0 when empty sender

0x346e3730 opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | N/A
| License       | MIT

Today I managed to produce the following error :
![image](https://github.com/symfony/symfony/assets/28731165/11b3b76d-d220-4f08-9b12-838ca936f540)

From the following code :
```php
$email = (new TemplatedEmail())
            ->from()
            ->to($member->getEmail())
            ->subject('Adhésion Sauvade')
            ->htmlTemplate('emails/new-membership/index.html.twig')
            ->context([
                'member' => $member,
            ])
            ->addPart(new DataPart(
                $this->pdfService->generatePdfFromTemplate(
                    'emails/new-membership/attachments/recu.html.twig',
                    ['member' => $member]
                ),
            ))
            ->addPart(new DataPart(
                $this->pdfService->generatePdfFromTemplate(
                    'emails/new-membership/attachments/membership-card.html.twig',
                    ['member' => $member]
                ),
            ))
        ;

        $this->mailer->send($email);
```

This happens because the `from` method from the `Symfony\Component\Mime\Email` class adds the header even if the array of addresses is empty. I identified two ways of fixing this :
- Adding a check in the `Symfony\Component\Mime\Message` class before accessing the key 0 and throw a better exception in case the adresses are empty
- Modifying the `from` method so that it doesn't add the header if the adresses are empty; or throw an exception when trying to do so.

I'm unsure what method is the cleaner, I thought that fixing the Message access of an undefined key was the widest fix (as I guess it could happen from somewhere else than the Mailer component). Also not sure how to test that and wanted to make sure this fix was good before testing it.

0x346e3730 committed:
[Mime] Fix undefined array key 0 when empty sender

xabbuh commented:
Can you please add a test to make sure that we don’t break this again.

0x346e3730 commented:
Done. It seems to me that the failing tests are related to cryptography errors and not my code

xabbuh commented on src/Symfony/Component/Mime/Tests/MessageTest.php:
```diff
@@ -125,6 +125,14 @@ public function testGetPreparedHeadersHasSenderWhenNeeded()
 
+    public function testGenerateMessageIdThrowsWhenHasFromButNoAddresses()
+    {
+        $this->expectException(\LogicException::class);
```
please move this line right before the call of `generateMessageId()`, also we should change the class name to actually check for the `LogicException` that is thrown in the `Message` class (i.e. the one from the Mime component)

xabbuh commented on src/Symfony/Component/Mime/Message.php:
```diff
@@ -140,7 +140,10 @@ public function generateMessageId(): string
             $sender = $this->headers->get('Sender')->getAddress();
         } elseif ($this->headers->has('From')) {
-            $sender = $this->headers->get('From')->getAddresses()[0];
+            if (empty($froms = $this->headers->get('From')->getAddresses())) {
```
```suggestion
            if ([] === $froms = $this->headers->get('From')->getAddresses()) {
```

0x346e3730 replied:
Used the proposition of Fabien instead, letting you resolve the conversation if it's okay with you

xabbuh commented on src/Symfony/Component/Mime/Message.php:
```diff
@@ -140,7 +140,10 @@ public function generateMessageId(): string
         } elseif ($this->headers->has('From')) {
-            $sender = $this->headers->get('From')->getAddresses()[0];
+            if (empty($froms = $this->headers->get('From')->getAddresses())) {
+                throw new LogicException('Your email has a From header that contains no addresses. Did you forgot to add emails in the "from" method ?');
```
```suggestion
                throw new LogicException('A "From" header must have at least one email address.');
```

fabpot commented:
What about forbidding calling `Email::from()` without any address instead?

fabpot commented on src/Symfony/Component/Mime/Message.php:
```diff
@@ -140,7 +140,10 @@ public function generateMessageId(): string
             $sender = $this->headers->get('Sender')->getAddress();
         } elseif ($this->headers->has('From')) {
-            $sender = $this->headers->get('From')->getAddresses()[0];
+            if (empty($froms = $this->headers->get('From')->getAddresses())) {
```
```suggestion
            if (!$froms = $this->headers->get('From')->getAddresses()) {
```

fabpot commented:
I would do the fix in `Email::from()` anyway as there is no good use case for not passing anything to this method.

0x346e3730 commented:
I adjusted the code with your reviews and added an exception if calling Email::from() with no addres, however the `from` method just calls `setListAddressHeaderBody` that is also called by `replyTo`, `to`, `cc`, `bcc` and `addListAddressHeaderBody`. Should we implement the throw in Email::setListAddressHeaderBody() instead so that it covers all method that add email addresses to certain headers ?

Plus the "add" version of those methods (`addFrom`, `addReplyTo` ect...) call `addListAddressHeaderBody` that does `$header->addAddresses(Address::createArray($addresses));`. Should we do the same throw in `MailboxListHeader::addAddresses` as well ?

xabbuh commented on src/Symfony/Component/Mime/Email.php:
```diff
@@ -121,6 +121,10 @@ public function addFrom(...$addresses)
     public function from(...$addresses)
     {
+        if (!$addresses) {
+            throw new LogicException(sprintf('"from()" must be called with at least one address.'));
```
```suggestion
            throw new LogicException('"from()" must be called with at least one address.');
```

fabpot commented:
Thank you @0x346e3730.
